akka-http-core/src/main/scala/akka/http/Http.scala

```diff
@@ -128,44 +135,44 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
                          log: LoggingAdapter = system.log): Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]] = {
     val effectiveSettings = ClientConnectionSettings(settings)
     val remoteAddr = new InetSocketAddress(host, port)
+    val layer = clientLayer(remoteAddr, effectiveSettings, log)
+
     val transportFlow = StreamTcp().outgoingConnection(remoteAddr, localAddress,
       options, effectiveSettings.connectingTimeout, effectiveSettings.idleTimeout)
-    val clientBluePrint = HttpClient.clientBlueprint(remoteAddr, effectiveSettings, log)
-
-    Flow(transportFlow, clientBluePrint)(Keep.left) { implicit b ⇒
-      (tcp, client) ⇒
-        import FlowGraph.Implicits._
 
-        tcp.outlet ~> client.bytesIn
-        client.bytesOut ~> tcp.inlet
-
-        (client.httpRequests, client.httpResponses)
-    }.mapMaterialized { tcpConnFuture ⇒
+    layer.joinMat(transportFlow) { (_, tcpConnFuture) ⇒
       import system.dispatcher
-      tcpConnFuture.map { tcpConn ⇒ OutgoingConnection(tcpConn.localAddress, tcpConn.remoteAddress) }
+      tcpConnFuture map { tcpConn ⇒ OutgoingConnection(tcpConn.localAddress, tcpConn.remoteAddress) }
     }
-
   }
 
   /**
-   * Transforms the given low-level TCP client transport [[Flow]] into a higher-level HTTP client flow.
+   * The type of the client-side HTTP layer as a stand-alone BidiStage
+   * that can be put atop the TCP layer to form an HTTP client.
+   *
+   * {{{
+   *                +------+
+   * HttpRequest  ~>|      |~> ByteString
+   *                | bidi |
+   * HttpResponse <~|      |<~ ByteString
+   *                +------+
+   * }}}
    */
-  def transportToConnectionClientFlow[Mat](transport: Flow[ByteString, ByteString, Mat],
-                                           remoteAddress: InetSocketAddress, // TODO: removed after #16168 is cleared
-                                           settings: Option[ClientConnectionSettings] = None,
-                                           log: LoggingAdapter = system.log): Flow[HttpRequest, HttpResponse, Mat] = {
-    val effectiveSettings = ClientConnectionSettings(settings)
-    val clientBlueprint = HttpClient.clientBlueprint(remoteAddress, effectiveSettings, log)
+  type ClientLayer = BidiFlow[HttpRequest, ByteString, ByteString, HttpResponse, Any]
```

Comment:
this is a return type that we provide to users, so it should be the concrete type that the blueprint materializes to